File: Src/iirSOS.c
/*
 * iirSOS.c
 *
 *  Created on: May 9, 2018
 *      Author: Ocanath
 */
#include "iirSOS.h"

//int sos(iirSOS * f, int newSample)
//{
//	f->w[2] = f->w[1];
//	f->w[1] = f->w[0];
//	f->w[0] = newSample - f->a1*f->w[1] - f->a2*f->w[2];
//	return f->b0 * f->w[0] + f->b1*f->w[1] + f->b2*f->w[2];
//}
float sos_f(iirSOS * f, float newSample)
{
	f->w[2] = f->w[1];
	f->w[1] = f->w[0];
	f->w[0] = newSample - f->a1*f->w[1] - f->a2*f->w[2];
	return f->gain*(f->b0 * f->w[0] + f->b1*f->w[1] + f->b2*f->w[2]);
}

File: MATLAB/iirSOS.m
function [y,w] = iirSOS(w,b0,b1,b2,gain,a1,a2, x)
    w(3) = w(2);
    w(2) = w(1);
	w(1) = x - a1*w(2) - a2*w(3);
    y = gain*(b0*w(1) + b1*w(2) + b2*w(3));   
end
